src/versionbits.h

```diff
@@ -52,17 +54,22 @@ class AbstractThresholdConditionChecker {
 
 public:
     // Note that the function below takes a pindexPrev as input: they compute information for block B based on its parent.
-    ThresholdState GetStateFor(const CBlockIndex* pindexPrev, const Consensus::Params& params, ThresholdConditionCache& cache) const;
+    ThresholdState CachedGetStateFor(const CBlockIndex* pindexPrev, const Consensus::Params& params, ThresholdConditionCache& cache) const;
 };
 
-struct VersionBitsCache
+class VersionBitsCache
 {
+private:
+    CCriticalSection cs;
```

Comment:
NACK on this; this class will be needed inside abstracted consensus logic, which shouldn't do its own logic.

VersionBitsCache is intended to be a black box data type locked by the user (main, in this case) but only accessed through functions in versionbits.cpp.